src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,175 +575,216 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
+      // If the state is invalid, the blocks must be deleted from the cache rather than written to the backend.
+      if (invalid) {
+	ldpp_dout(dpp, 10) << __func__ << "(): State is INVALID; deleting object." << dendl;
+	int ret = -1;
+	if ((ret = cacheDriver->delete_data(dpp, "D#" + e->key, y)) == 0) { // Sam: do we want del or delete_data here?
+	  if (!(ret = erase(dpp, e->key, y))) {
+	    ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << e->key << ", ret=" << ret << dendl;
+	    continue;
+	  }
+	} else {
+	  ldpp_dout(dpp, 0) << "Failed to delete head object for: " << e->key << ", ret=" << ret << dendl;
+	  continue;
+	}
 
-      int op_ret = processor->prepare(null_yield);
-      if (op_ret < 0) {
-	ldpp_dout(dpp, 20) << __func__ << "processor->prepare() returned ret=" << op_ret << dendl;
-        erase_dirty_object(dpp, e->key, null_yield);
+	ret = delete_data_blocks(dpp, e, y);
+	if (ret == 0) {
+	  erase_dirty_object(dpp, e->key, null_yield);
+	} else {
+	  ldpp_dout(dpp, 0) << "Failed to delete blocks for: " << e->key << ", ret=" << ret << dendl;
+	  continue;
+	}
       }
 
-      std::string prefix = url_encode(e->bucket_id) + CACHE_DELIM + url_encode(e->version) + CACHE_DELIM + url_encode(c_obj->get_name());
-      off_t lst = e->size;
-      off_t fst = 0;
-      off_t ofs = 0;
-
-      rgw::sal::DataProcessor *filter = processor.get();
-      std::string head_oid_in_cache = DIRTY_BLOCK_PREFIX + prefix;
-      std::string new_head_oid_in_cache = prefix;
-      ldpp_dout(dpp, 10) << __func__ << "(): head_oid_in_cache=" << head_oid_in_cache << dendl;
-      ldpp_dout(dpp, 10) << __func__ << "(): new_head_oid_in_cache=" << new_head_oid_in_cache << dendl;
-      bufferlist bl;
-      cacheDriver->get_attrs(dpp, head_oid_in_cache, obj_attrs, null_yield); //get obj attrs from head
-      obj_attrs.erase(RGW_CACHE_ATTR_MTIME);
-      obj_attrs.erase(RGW_CACHE_ATTR_OBJECT_SIZE);
-      obj_attrs.erase(RGW_CACHE_ATTR_ACCOUNTED_SIZE);
-      obj_attrs.erase(RGW_CACHE_ATTR_EPOCH);
-
-      do {
-        ceph::bufferlist data;
-        if (fst >= lst){
-	  break;
-        }
-        off_t cur_size = std::min<off_t>(fst + dpp->get_cct()->_conf->rgw_max_chunk_size, lst);
-        off_t cur_len = cur_size - fst;
-        std::string oid_in_cache = DIRTY_BLOCK_PREFIX + prefix + CACHE_DELIM + std::to_string(fst) + CACHE_DELIM + std::to_string(cur_len);
-        ldpp_dout(dpp, 10) << __func__ << "(): oid_in_cache=" << oid_in_cache << dendl;
-        rgw::sal::Attrs attrs;
-        cacheDriver->get(dpp, oid_in_cache, 0, cur_len, data, attrs, null_yield);
-        len = data.length();
-        fst += len;
-
-        if (len == 0) {
-          // TODO: if len of any block is 0 for some reason, we must return from here?
-          break;
-        }
+      if (!invalid && (std::difftime(time(NULL), e->creationTime) > interval)) { // if block is dirty and written more than interval seconds ago
```

Comment:
is the check for !invalid really needed here?

Reply:
It's possible for the state to be invalid and for an interval amount of seconds to have passed since the entry's creation, so I think it is necessary to ensure the cleaning method doesn't attempt to write the object to the backend after it has already been deleted.